Enable monthly freq for cost_gencost_sshv4

antnguyen13 opened the pull request:
## What changes does this PR introduce?
Enable switch from reading/processing only daily to also reading/processing monthly data and costs in pkg/ecco ssh cost.


## What is the current behaviour? 
Currently the pkg/ecco ssh cost (e.g., cost_gencost_sshv4) can only read in daily data, with hardcoded numbers such as 35 days, 365 or 366 days.  This cost has very high I/O load, for example, when reading in three to four daily data sets, we're required to open and close multiple files nearly simultaneously and continuously as well as writing out large files (~25yr * 365day/yr * nx * ny * 3 products) and read in again to compute mean/smooth/misfits.  For high resolution models, this I/O load can be too high such that the model will crash and as a result ssh daily cost can not be part of the cost function for an optimization.


## What is the new behaviour 
An introduction of logical use_mon and appropriate code modification to enable the switch from daily to monthly.  This allows users to read in monthly data and still have misfits for ssh, and subsequently significantly lowers the I/O load.  There is a consequence that the high (daily) frequency variability can be missing when reading in monthly data.  However, this capability still allows the model to have monthly mean ssh close to observations and ssh misfits to be part of the cost function.   
In addition to the main modification, I made one more modification to check the range of the years for MDT calculations to not only be based on some hardcoded MDT [start,end] date but also on the model [start,end] dates, to allow for cases outside of eccov4 where the startdate is not necessarily 1992 but overlap enough with the MDT period that MDT can still be used for misfit.


## Does this PR introduce a breaking change? 
I don't expect any changes.  I have switched between daily and monthly, with daily still reproducing exact existing code results.  


## Other information:
For my configuration ASTE, on certain HPC such as TACC, daily ssh cost creates 100% failure due to high load of I/O .  The switch to monthly allows me to continue to have cost as constraints even at lower temporal resolution.  For my higher res run based on llc1080, the daily ssh file size would be enormous and the I/O load prohibitive.  This switch enables cost to ssh to be computed.

## Suggested addition to `tag-index`

timothyas commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -268,16 +277,44 @@ subroutine cost_gencost_sshv4(
-c-- so for short model run, this will always get set to FALSE
+c-- but that there has to be at least 365 days (or 12 months) of model
+c-- run so for short model run, this will always get set to FALSE
+catn: note also the hard-coded years 1992,2011 here that SHOULD BE REMOVED
```
Why not delete it? Will this cause problems for ECCO?

antnguyen13 replied:
@timothyas I didn't touch these hardcoded numbers but have talked to @owang01 to take a look.  I'm fairly certain the current eccov4r* mdt date range is beyond 1992-2011, so I don't know why yet eccov4 still checking this date range.. But I think it should be addressed at a later time perhaps, and not part of this PR.

timothyas commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -211,11 +214,17 @@ subroutine cost_gencost_sshv4(
        do k2 = 1, NGENPPROC
         if (gencost_posproc(k2,igen_mdt).EQ.'smooth') k2_mdt=k2
        enddo
+       write(msgbuf,'(a,i5)') ' sshv4: k2_mdt: ',k2_mdt
+       call print_message( msgbuf, standardmessageunit,
+     &                     SQUEEZE_RIGHT , myThid)
```
What do you think about printing this `#ifdef ECCO_VERBOSE`? If you agree, I'd suggest that for essentially all of the print_messages in this subroutine that print out loop counters.

antnguyen13 replied:
There are other print_message I've added in the past, as well as many added by others to help the users understand the parameters being set for this cost here (e.g., mdt, lsc, which are dependent on some setting within this subroutine).  Based on my experience, this can be a *very* confusing subroutine with cost(s) missing/being turned off without much explanation and it takes a lot of time to find out why.  So I have put some comments into ECCO_VERBOSE per your suggestion (see further below) but want to leave some quite essential ones as default to help the users.  I've modified this to also print out k2_lsc and is part of default printing.

timothyas replied:
That sounds good to me, thanks!

timothyas commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -415,7 +460,8 @@ subroutine cost_gencost_sshv4(
 
       do year=mdt_y0,mdt_y1
-       do day=1,366
+c       do day=1,366
```
I'd suggest to delete this rather than comment it out

antnguyen13 replied:
thanks. done

timothyas reviewed (commented):
Thanks for this @antnguyen13! I have never used this subroutine and find it pretty complicated and intimidating, so my comments are somewhat basic, but either way this is a great improvement. It would definitely good to get @owang01's eyes on this since it's ECCO specific.

antnguyen13 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -591,19 +640,33 @@ subroutine cost_gencost_sshv4(
       offset_sum = 0. _d 0
 
-      do irec = 1, ndaysrec
+c      write(msgbuf,'(a,i)')
+c     & 'sshv4: before entering irec loop, nmonsrec:',nmonsrec
+c      call print_message( msgbuf, standardmessageunit,
+c     &                    SQUEEZE_RIGHT , myThid)
+c      write(msgbuf,'(a,i)')
+c     & 'sshv4: before entering irec loop, ndaysrec:',ndaysrec
+c      call print_message( msgbuf, standardmessageunit,
+c     &                    SQUEEZE_RIGHT , myThid)
+c      write(msgbuf,'(a,i)')
+c     & 'sshv4: before entering irec loop, nrec_total:',
+c     &    nrec_total
+c      call print_message( msgbuf, standardmessageunit,
+c     &                    SQUEEZE_RIGHT , myThid)
+catn      do irec = 1, ndaysrec
```
I've consolidated these 3 prints into one but will leave as default.

antnguyen13 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -993,21 +1076,22 @@ subroutine cost_gencost_sshv4(
 c ------------------------------------------------------
 
-      do jrec=1,ndaysave
+catn      do jrec=1,ndaysave
```
done.

antnguyen13 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -973,7 +1055,8 @@ subroutine cost_gencost_sshv4(
 catn -----------
 
-      do irec = 1, ndaysrec-ndaysave+1
+catn      do irec = 1, ndaysrec-ndaysave+1
```
done

antnguyen13 commented on pkg/ecco/cost_sla_read_yd.F:
```diff
@@ -1,11 +1,11 @@
-     I                sla_file, sla_startdate,
+     I                sla_file, sla_startdate, use_mon,
+     O                sla_obs, sla_mask,
+     I                year, irec, myThid )
```
jeah, i've deleted them. thanks.

antnguyen13 commented on pkg/ecco/cost_sla_read_yd.F:
```diff
@@ -78,21 +82,33 @@ subroutine cost_sla_read_yd(
+catn
+      read_rec=.FALSE.
+      if (use_mon) then
+        if ( (irec.GE.1).AND.(irec.LE.12 ) ) then
```
I don't think we need to print the warning in here because the irec can only loop between 1 to 12 when use_mon is .true. , after going through a series of tests in cost_gencost_sshv4.F.  Specifically, as default, use_mon = .false., and irec goes from 1:365 or 1:366.  ONLY when gencost_avgperiod = "mon" that use_mon is set to .true., and the range of irec is being forced to be within 1:12 only.  The flag use_mon as well as the range of irec are then immediately print out to STDOUT .  Thus, there is no circumstance in which use_mon is .true. and irec can be outside of 1:12 as far as i can see, unless you spot it somewhere?

timothyas replied:
Ah ok, thanks I see that now. I think it's OK without the warning.

owang01 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -68,14 +68,17 @@ subroutine cost_gencost_sshv4(
       integer igen_mdt, igen_lsc, igen_gmsl
       integer igen_tp, igen_ers, igen_gfo
-      integer igen_etaday
+      integer igen_etaobs
```
Using etaobs is a bit confusing, as the field is actually the model equivalent of SSH. So it is probably better to use etaps (to be consistent with psmean etc. later in the S/R) or etamd.

antnguyen13 replied:
We have discussed, this will now be called etagcm.

owang01 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -99,7 +102,7 @@ subroutine cost_gencost_sshv4(
       _RL ersma(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
       _RL gfoma(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
-      _RL etaday(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
+      _RL etaobs(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
```
Same as the comment above; use etaps

antnguyen13 replied:
now etagcm. done.

owang01 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -948,20 +1006,41 @@ subroutine cost_gencost_sshv4(
+      endif
+      nrecsaveRL=nrecsave
 
 catn add a warning if have less nrecday than the hardcoded 35days
```
catn add a warning if have less nrec_total than the hardcoded nrecsave

antnguyen13 replied:
thanks. done.

owang01 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -1472,7 +1557,8 @@ subroutine cost_gencost_sshv4(
 
-      enddo !do irec = 1, ndaysrec-ndaysave+1
+catn: closing irec loop below
+      enddo
```
Add back the comment about which the loop it is
`      enddo !do irec = 1, nrec_total-nrecsave+1`

antnguyen13 replied:
thanks. done put back.

owang01 commented on pkg/ecco/cost_sla_read.F:
```diff
@@ -85,6 +95,25 @@ subroutine cost_sla_read(
+
+      else
+
+        dayspermon = 1.0
```
`dayspermon = 1.0` for daily means is confusing. Should `dayspermon` be replaced by `daysperrec` instead?

antnguyen13 replied:
I've changed dayspermon to daysperavgperiod to make it clearer.
I've also gone ahead and change "daydate", "dayiter", "daytime" to "obsdate", "obsiter", "obstime", and "yday" to "yobs" to help reducing confusion related to the originally hardcoded "day" that was meant to be daily observation.

owang01 commented on pkg/ecco/cost_sla_read.F:
```diff
@@ -103,6 +132,7 @@ subroutine cost_sla_read(
       call cal_ToSeconds( difftime, diffsecs, myThid )
 c      sshrec = floor(diffsecs/sla_period) + 1
       sshrec = int(diffsecs/sla_period) + 1
```
Adding two extra spaces in the beginning of each line to increase indentation.

antnguyen13 replied:
thanks. Done. i also change several other lines' indentation to make the whole thing easier to read.

owang01 commented on pkg/ecco/cost_sla_read_yd.F:
```diff
@@ -78,21 +80,33 @@ subroutine cost_sla_read_yd(
+      else
+
+        if ( (irec.GE.1).AND.( (
+     &   (cal_IsLeap(year,myThid).eq.2).AND.(irec.LE.366)
```
Change `eq` to `EQ`

antnguyen13 replied:
done.

owang01 commented on pkg/ecco/cost_sla_read_yd.F:
```diff
@@ -78,21 +80,33 @@ subroutine cost_sla_read_yd(
       CALL READ_REC_3D_RL( fnametmp, cost_iprec, 1,
-     &                   sla_obs, day, 1, myThid )
+     &                   sla_obs, irec, 1, myThid )
 
```
Increase indentation for the lines from `CALL READ_REC_3D_RL` through the end of the IF block. Capitalize `eq`, `le`, `lt` at lines 117, 122, and 125.

antnguyen13 replied:
done. i checked the entire subroutine to get spacing / indentation consistent, and also changed all the if , else, etc, to all caps.

owang01 commented on pkg/ecco/cost_sla_read_yd.F:
```diff
@@ -132,7 +146,7 @@ subroutine cost_sla_read_yd(
 
-      else !if ( (day.GE.1).AND...
+      else !if ( (irec.GE.1).AND...
 
```
Increase indentation for the lines in the else block.

antnguyen13 replied:
done.

owang01 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -948,20 +1006,41 @@ subroutine cost_gencost_sshv4(
+catn: nrecsave is used for running mean, so should be > 1
+catn: also, nrecsave/2 is used as time of middle running-mean period
+catn: so it MUST BE AT LEAST 2 or else large-scale cost WILL BE SKIPPED!!
+        nrecsave=2
```
Why `nrecsave=2`? For daily means, the S/R applies 35 days moving average for the lsc SSH cost. For monthly means, do we still need to apply the temporal moving average? Setting `nrecsave=1` for monthly means is more consistent with 35 day moving average for daily means. `nrecsave=2` would imply ~60 day moving average.

antnguyen13 replied:
nrecsave is now set to 1.

owang01 reviewed (changes requested):
@antnguyen13 Thank you for the PR. Except for a few minor comments, overall it looks good to me.

antnguyen13 commented:
@owang01 Thanks for your review again, i've made all the changes suggested by you (etaobs -> etagcm, spacing, indentation to improve readability).  I also renamed one variable to  "daysperavgperiod"  to really reflect the meaning of it.

@jm-c : while i was editing the file cost_gencost_sshv4.F , the file indentation was a hassle because there are many loops and i couldn't track well due to spacing all over the place.  I could also comb through to get the spacing all consistent, if you think that's a good idea? I did it, then undid it.. because it looked as if I changed the entire file (it will look that way, since nearly the entire routine requires re-spacing..)

If we don't need spacing/indentation mod, everything is good to go on my end, as soon as it passes whatever else it needs to pass before getting merged.

jm-c reviewed (commented):
I think this PR is a valuable contribution. But as I don't know much about this ECCO cost specific part, I could easily miss something. I am currently running `testreport -adm` in the `verification_other` set of experiments to double-check that every tests pass (I already did it in verification and all was good).

I have few comments, mostly related to ... comments:

1. The argument "use_mon" has been added to the 2 routines `cost_sla_read.F` and `cost_sla_read_yd.F`. But in each file, it's declared within the set of local variables instead of as a new argument. And I think it would be very useful also to add 1 line of description for this new argument. I know the other arguments are not described, but I still think it's would be useful, at least for this new one.
2. There are few new un-used variables that could be safely removed. I get a warning about: "day" in `cost_gencost_sshv4.F` and about "beginrun", "beginssh", "nobs" in `cost_sla_read.F`, something to check.
3. I think it would be useful to remind how to switch from daily frequency to monthly frequency (this PR goal) as a user point of view. This could be mentioned in the PR description and also somewhere, as a comment, in `cost_gencost_sshv4.F`. Now in `ecco_readparms.F`, `ecco_cost_init_fixed.F` & `cost_gencost_assignperiod.F` it seems that both "month" and "MONTH" are recognized. But here, lines 284-286, we only check for the first 3 characters and only lower-case. Is there a good reason for that ?
4. There are some good additional comments in the code, but not sure we want to keep all of them (see few comments below) and would be better without starting "catn " but instead simply "C ".

jm-c commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -268,16 +274,47 @@ subroutine cost_gencost_sshv4(
       call print_message( msgbuf, standardmessageunit,
      &                    SQUEEZE_RIGHT , myThid)
 
+catn: loop once to determine use_mon and nrec_year
```
Also, I don't know what this "loop once" means exactly, since there is no following loop. And not sure neither about the following 2 lines of comment.

antnguyen13 replied:
I've modified lines 277-279 (3 lines of unclear comments) to:

C Check to determine use_mon and nrec_year.  Check all igen_[tp,ers,gfo]
C in case some of the data files are not defined.

In addition, at the top of the file when use_mon is defined, i also add a comment above it:

c introduce use_mon to switch from using daily to monthly data if true

antnguyen13 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -494,7 +545,7 @@ subroutine cost_gencost_sshv4(
        endif
 
-       enddo !do day=1,366
+       enddo !do day=1,366 or irec=1,12
```
I've changed this now to !do irec=1,nrec_year

antnguyen13 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -268,16 +274,47 @@ subroutine cost_gencost_sshv4(
 c minimum number of observations to consider re-referenced MDT
-      num0=100
+      IF (use_mon) THEN
+catn: should set to 3 and 12
```
I have removed this catn comment line altogether.

antnguyen13 commented on pkg/ecco/cost_gencost_sshv4.F:
```diff
@@ -68,14 +68,17 @@ subroutine cost_gencost_sshv4(
       integer imin,imax
       integer irec,jrec,krec
       integer ilps
+catn
```
line has been removed.

antnguyen13 commented:
@jm-c Thank you for the detailed review. I think i've addressed all of your comments in my latest commit, most noticeable:

- I added, in each of the files cost_gencost_sshv4.F, cost_sla_read.F, cost_sla_read_yr.F, 1 line for explaining the new logical variable use_mon
- I added additional check for "MON" and not just "mon" to trigger the setting for use_mon in cost_gencost_sshv4.F
- I cleaned up all the catn comments, changed to "C" or "c" and modified them when needed to try to make things clearer
- I removed unused variables and move use_mon into the "argument variables" instead of in "local variables" section.

I will now update the PR description to reflect clearer the goal of the PR, enable switching from daily to monthly.

jm-c commented:
@antnguyen13 Regarding my previous question (point 3):
> Now in ecco_readparms.F, ecco_cost_init_fixed.F & cost_gencost_assignperiod.F it seems that both "month" and "MONTH" are recognized. But here, lines 284-286,
>  we only check for the first 3 characters and only lower-case. Is there a good reason for that ?

Could you explain why we only check for the first 3 characters ?

antnguyen13 commented:
oh , sorry @jm-c I believe in my latest mod , the check is now for both "mon" and "MON", did you see it? (I had forgotten to say it in my commit)

jm-c commented:
I found my remaining comments somewhere, and it's about the setting of the new switch "use_mon" in `cost_gencost_sshv4.F` (lines 277-293):

1. when one or two of the 3 costs is used but not not all 3, I will have one (at least) of "igen_[tp,ers,gfo]" that will be non-zero and one (at least) that will be zero. This will trigger an index out-of-bounds for the zero value in the corresponding test (lines 283-288).
2. the "use_mon" applies to all 3 costs (tp,ers,gfo), but as soon as one of the 3 is set to monthly averaged, the 2 other will be switched also even if the user did not specify a corresponding `gencost_avgperiod` to be "month" for the 2 others. Do we need  a different "use_mon" for each of the 3 ? or may-be, if this is an unlikely situation, a stop somewhere (in `ecco_check.F`) to prevent this to happen ?

jm-c commented on pkg/ecco/cost_sla_read.F:
```diff
@@ -41,6 +43,7 @@ subroutine cost_sla_read(
       _RL sla_mask(1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
       integer irec
       integer myThid
+      logical use_mon
```
All the other arguments are declared in the same order as they are in the argument list (was done in PR #579).

antnguyen13 replied:
this is now fixed.

jm-c commented on pkg/ecco/cost_sla_read_yd.F:
```diff
@@ -40,8 +40,9 @@ subroutine cost_sla_read_yd(
-      integer year, day
+      integer year, irec
       integer myThid
+      logical use_mon
```
same comment here.

antnguyen13 replied:
this is now fixed.

jm-c commented:
@antnguyen13 Thanks for the very detailed and useful reply comments. I agree with this. Will take a look soon to the latest changes.

jm-c commented:
@antnguyen13 Could you check the latest changes in `ecco_check.F` ? On my side, it fails to compile.

jm-c commented:
Still unable to compile "ecco_check.F"  (I tried in experiment "lab_sea", Adjoint built).

antnguyen13 commented:
@jmc: oh, hmm, let me check!

antnguyen13 commented on pkg/ecco/cost_sla_read.F:
```diff
@@ -35,6 +37,7 @@ subroutine cost_sla_read(
       character*(MAX_LEN_FNAM) sla_file
       integer sla_startdate(4)
       _RL sla_period
+      logical use_mon
```
need 1 line to description

jm-c commented:
I am going to push few little changes: 
a) in `ecco_check.F`, just to match exactly how "use_mon" is set in `cost_gencost_sshv4.F`, I can simplify `use_day_int` by counting any of the 3 that are not monthly.
b) in `cost_sla_read.F`, "beginmodel" and "mody" are the same, also "beginlocal"  and "ymod" are the same,
and "obsrec" is set but never used. Would like to remove all these 3 local vars.

jm-c commented:
@antnguyen13 I put the changes in `cost_sla_read.F` in a separated commit (my last one) so that it could be reverted if it is not right. Please check what I did. 
Otherwise, from side, this would be good to be merged once you agree.

antnguyen13 commented:
@jm-c From my computer, i originally did a git pull origin cost_sshv4_month to grab your changes from 2 days ago [bf5d283](https://github.com/MITgcm/MITgcm/pull/694/commits/bf5d2834246a9a180b9c8edf84815a3287ffb22b), and did a test, all results are the same as my commit just before that [c9da9b2](https://github.com/MITgcm/MITgcm/pull/694/commits/c9da9b2678103a019acd06562125a03a16b47f73) .  So based on that, everything LGTM.

A note: When i came here to leave this comment, i got a "must resolve 3 conflicts" .. I think it is related to the 80char PR that got merged in ? so I did try to merge, could you check if my latest commit related to these files w/ 1:80 being removed as correct?  This is the first time i've encountered this forced resolve on the git website..

jm-c commented:
@antnguyen13 Update: changes were made (from PR #715) to these 3 files: `cost_gencost_sshv4.F` and `cost_sla_read[_yd].F` but the merging into your branch was done cleanly (you did it very well !) so the resulting updated branch looks good.

jm-c commented:
Will merge this PR soon, like tomorrow, unless someone wants to add something.
